### ISSUE TITLE
[FEATURE REQUEST] Provide some kind of functionality to help with providing C bindings for a library

### ISSUE THREAD

**Issue by ids1024** [gtk-rs/gtk-rs-core#724]
As far as I'm aware (I really need to experiment with it more) to create a library in gtk-rs that could be used from other languages (C, Vala, Python, C++, etc.), or from Rust but dynamically with a stable ABI, it would be necessary to:

* Write unsafe `extern "C"` bindings around any methods that should be exposed.
* Manually write a `.h` header, with necessary `transfer`, `nullable`, etc. annotations.
* Run `g-ir-scanner` to generate metadata from the header.

This isn't terrible, but I wonder if it would be possible to offer something for this that's kind of the opposite of the `gir` crate, to help generate the C wrappers and header:

* Some sort of proc macro to annotate functions, to generate a C wrapper while writing just safe Rust code using native Rust types.
* Automatically generate C headers matching the binding, with the correct `transfer` and `nullable` annotations.
* Extract rustdoc documentation, and include in header file too.

Ideally this would provide a way to dynamically link a library built in gtk-rs from an application build in gtk-rs without writing anything unsafe on either end. And the library could be used from other languages.

**Comment by bilelmoussaoui** [comment]
> Oh, and I didn't even mention that it would need to expose a _get_type() function, a #define calling that, and anything else needed for binding a GObject.

The _get_type is already exposed through the StaticType trait. I would say, we need a new macro or some attribute for each macro that would generate the extra traits needed for providing bindings.

**Comment by ids1024** [comment]
Yeah, `StaticType` already covers that, it would just need a `#[no_mangle] extern "C"` wrapper to be called from outside Rust, which I don't think there is currently. But that's simple enough.

**Comment by ids1024** [comment]
So something like seems to work:

```rust
#[no_mangle]
extern "C" fn my_type_get_type() -> glib::ffi::GType {
    MyType::static_type().into_glib()
}
```

Plus a header file with the usual boilerplate. The hard part is if you want a lot of methods, and need C bindings for those. What could be handy is some kind of proc macros for defining C ABI functions wrapping Rust types. A flexible enough solution would also be helpful for defining trampoline functions for callbacks, and for methods when defining `IsSubclassable`.

This seems problematic with gtk3-rs/gtk4-rs due to the `INITIALIZED` logic, since the caller of the cdynlib won't be setting that even if it has called `gtk_init`.

**Comment by sdroege** [comment]
> This seems problematic with gtk3-rs/gtk4-rs due to the `INITIALIZED` logic, since the caller of the cdynlib won't be setting that even if it has called `gtk_init`.

https://gtk-rs.org/gtk3-rs/stable/latest/docs/gtk/fn.set_initialized.html

**Comment by ids1024** [comment]
It's not clear where to call that without the library having it's own `init` function, which seems like it shouldn't be necessary for every C library providing widgets.

I ran into this just calling `g-ir-scanner`. I don't think calling `static_type()` on a `GtkWidget` subclass should actually require that GTK is initialized, but seems to be checking that somewhere. And trying to call `set_initialized` in the `*_get_type()` function errors with "GTK was not actually initialized". Calling `gtk4::init()` there works but seems like it's definitely the wrong solution.

**Comment by jf2048** [comment]
I have a plan for doing this using the class macro: https://github.com/jf2048/gobject/issues/28
